File: 2216-delete-the-middle-node-of-a-linked-list/delete-the-middle-node-of-a-linked-list.py
class Solution(object):
    def deleteMiddle(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        # Case 1: Only one node → delete it
        if head is None or head.next is None:
            return None

        fast, slow = head, head
        prev = None

        # Step 1: Find middle with slow/fast pointers
        while fast and fast.next:
            fast = fast.next.next
            prev = slow
            slow = slow.next

        # Step 2: Delete middle node
        prev.next = slow.next

        return head

File: 2216-delete-the-middle-node-of-a-linked-list/delete-the-middle-node-of-a-linked-list.cpp
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        // Case 1: 0 or 1 node
        if (head == nullptr || head->next == nullptr)
            return nullptr;

        ListNode* fast = head;
        ListNode* slow = head;
        ListNode* prev = nullptr;

        // Find middle node
        while (fast != nullptr && fast->next != nullptr) {
            fast = fast->next->next;
            prev = slow;
            slow = slow->next;
        }

        // Delete middle node
        prev->next = slow->next;

        return head;
    }
};
